contracts/multisig-prover/src/query.rs

```diff
@@ -48,3 +57,89 @@ pub fn get_proof(deps: Deps, multisig_session_id: Uint64) -> StdResult<GetProofR
 pub fn get_worker_set(deps: Deps) -> StdResult<WorkerSet> {
     CURRENT_WORKER_SET.load(deps.storage)
 }
+
+/// Returns a new vector of signers with signatures containing the minimum amount of signatures required to reach quorum.
+/// The vector is sorted by weight and then is iterated over, removing signatures after the quorum is reached.
+fn optimize_signers(
+    quorum: Uint256,
+    mut signers: Vec<(Signer, Option<Signature>)>,
+) -> Vec<(Signer, Option<Signature>)> {
+    signers.sort_by(|(a, _), (b, _)| b.weight.cmp(&a.weight));
+
+    signers
+        .into_iter()
+        .scan(
+            Uint256::zero(),
+            |acc, (signer, signature)| match signature {
+                Some(sig) if *acc < quorum => {
+                    *acc += signer.weight;
+                    Some((signer, Some(sig)))
+                }
+                _ => Some((signer, None)),
+            },
+        )
+        .collect()
+}
```

Comment:
Do we still need the signers without signatures? I suggest something along these lines. If signers are still needed at least use `sorted_by` so you can chain everything together

```suggestion
) -> Vec<(Signer, Signature)> {
        signers
            .into_iter()
            .filter_map(|(signer, sig)| {
                if let Some(sig) = sig {
                    Some((signer, sig))
                } else {
                    None
                }
            })
            .sorted_by(|(a, _), (b, _)| b.weight.cmp(&a.weight))
            .scan(Uint256::zero(), |acc, (signer, signature)| {
                if *acc < quorum {
                    *acc += signer.weight;
                    Some((signer, signature))
                } else {
                    None
                }
            })
            .collect()
    }
```

Reply:
thanks! I was looking for something like `.sorted_by` but couldn't find it. Signers without signatures are required though, the proof contains all possible participants.